packages/web-media-box/src/pipelines/mse/mseManager.ts

```diff
@@ -0,0 +1,178 @@
+            resolve();
+          };
+          const onError = (event: Event) => {
+            removeAllEventListeners();
+            this.logger.warn(`Error ${event} appending to SourceBuffer of type ${mimeType}`);
+            reject();
+          };
+          const onAbort = (event: Event) => {
+            removeAllEventListeners();
+            this.logger.debug(`Aborted ${event} appending to SourceBuffer of type ${mimeType}`);
+            reject();
+          }
+          buffer.addEventListener('updateend', updateEnd);
+          buffer.addEventListener('error', onError);
+          buffer.addEventListener('abort', onAbort);
+          try {
+            buffer.appendBuffer(data);
+          } catch (error) {
+            this.logger.warn(`${error} cannot append data for ${mimeType} to the SourceBuffer`);
+          }
+        });
+      }
+      const appendOperation: SourceBufferOperation = {
+        type: OperationType.APPEND,
+        operation
+      };
+      queue.push(appendOperation);
+      this.processQueue(bufferWrapper);
+    }
+  }
+
+  public removeData(mimeType: string, start: number, end: number = Infinity) {
+    const bufferWrapper = this.sourceBuffers.get(mimeType);
+    if (bufferWrapper) {
+      const buffer = bufferWrapper.buffer;
+      const queue = bufferWrapper.queue;
+      const operation = (): Promise<void> => {
+        return new Promise((resolve, reject) => {
+          const removeAllEventListeners = () => {
+            buffer.removeEventListener('updateend', updateEnd);
+            buffer.removeEventListener('error', onError);
+            buffer.removeEventListener('abort', onAbort);
+          };
+          const updateEnd = () => {
+            removeAllEventListeners();
+            this.logger.debug(`Remove complete to ${mimeType} SourceBuffer from ${start} to ${end}`);
+            resolve();
+          };
+          const onError = (event: Event) => {
+            removeAllEventListeners();
+            this.logger.warn(`Error ${event} removing data from SourceBuffer of type ${mimeType}`);
+            reject();
+          };
+          const onAbort = (event: Event) => {
+            removeAllEventListeners();
+            this.logger.debug(`Aborted ${event} removing data from SourceBuffer of type ${mimeType}`);
+            reject();
+          }
+          buffer.addEventListener('updateend', updateEnd);
+          buffer.addEventListener('error', onError);
+          buffer.addEventListener('abort', onAbort);
+          try {
+            buffer.remove(start, end);
+          } catch (error) {
+            this.logger.warn(`${error} cannot remove data for ${mimeType} to the SourceBuffer from ${start} to ${end}`);
+          }
+        });
+      }
+      const appendOperation: SourceBufferOperation = {
+        type: OperationType.REMOVE,
+        operation
+      };
+      queue.push(appendOperation);
+      this.processQueue(bufferWrapper);
+    }
+  }
```

Comment:
seems like we can create a generic private method for this, like addOperation:

```ts
public appendData(mimeType, data): void {
return this.addOperation(mimeType, Operation.Append, (buffer) => buffer.appendData(data))
}

public removeData(mimeType, start, end): void {
return this.addOperation(mimeType, Operation.Remove, (buffer) => buffer.remove(start, end))
}

```

Reply:
done